### ISSUE TITLE
In 0.3.22 updates to documents are not always being communicated back to the client correctly

### ISSUE THREAD

**Issue by JackAdams** [arunoda/meteor-smart-collections#39]
There are still occasional discrepancies between the version of a document in the db and the version in minimongo.  Note: this only happens rarely, but it's confusing for the user when it does happen.  The exact scenario: an update is made to the db document and this update is communicated to all other connected clients, except the client that actually made the update. (It takes a browser refresh to see the update on the client that made the update.)

**Comment by arunoda** [comment]
Yes this is a really heavy issue. I will do another code review this weekend to see if I can find the actual issue. 

> Actually I was quite tired of this bug :)

**Comment by JackAdams** [comment]
Yeah, I appreciate it's a tough one -- non-deterministic.

Apart from this bug (which my users are being very accepting of) Smart Collections is working great and is the only reason we're able to use my Meteor app in production, serving our whole school.

Thanks for looking into it again.

Brent

**Comment by arunoda** [comment]
Wow. Nice.

I had some very busy month. I'll spend my full attention this in the this
weekend and the next week.

On Fri, Nov 8, 2013 at 11:08 AM, Brent Abrahams notifications@github.comwrote:

> Yeah, I appreciate it's a tough one -- non-deterministic.
> 
> Apart from this bug (which my users are being very accepting of) Smart
> Collections is working great and is the only reason we're able to use my
> Meteor app in production, serving our whole school.
> 
> Thanks for looking into it again.
> 
> Brent
> 
> ---
> 
> From: Arunoda Susiripala notifications@github.com
> To: arunoda/meteor-smart-collections <
> meteor-smart-collections@noreply.github.com>
> Cc: Brent Abrahams brent_abrahams@yahoo.com
> Sent: Friday, November 8, 2013 1:17 PM
> Subject: Re: [meteor-smart-collections] In 0.3.22 updates to documents are
> not always being communicated back to the client correctly (#39)
> 
> Yes this is a really heavy issue. I will do another code review this
> weekend to see if I can find the actual issue.
> Actually I was quite tired of this bug :)
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/arunoda/meteor-smart-collections/issues/39#issuecomment-28037089
> .

## 

Arunoda Susiripala

@arunoda http://twitter.com/arunoda
http://gplus.to/arunodahttps://github.com/arunoda
http://www.linkedin.com/in/arunoda

**Comment by arunoda** [comment]
Hi, sorry for the late update.

I fixed some few bugs in the invalidations. May be that cause the issues. Can you please try v0.3.24

**Comment by JackAdams** [comment]
Thanks for this. I'll try it out and let you know how it goes.  I've got a feeling it wasn't the invalidations though, it was the doc in the minimongo collection not matching the doc in the db collection.

**Comment by arunoda** [comment]
Yep,  than means this only occured if he is the person who is writing to
the db right? If this so this can be a fix.

Anyway I will keep looking into it.

On Saturday, December 7, 2013, Brent Abrahams wrote:

> Thanks for this. I'll try it out and let you know how it goes.  I've got a
> feeling it wasn't the invalidations though, it was the doc in the minimongo
> collection not matching the doc in the db collection.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/arunoda/meteor-smart-collections/issues/39#issuecomment-30043759
> .

## 

Arunoda Susiripala

@arunoda http://twitter.com/arunoda
http://gplus.to/arunodahttps://github.com/arunoda
http://www.linkedin.com/in/arunoda

**Comment by JackAdams** [comment]
Updated to 0.3.24 and we're still noticing this quite regularly -- where the minimongo colllection is updated during latency compensation, the database collection is updated correctly on the server, then when the ddp result comes back to the client who initiated the change, the changes are undone on the client (net result as a user -- you see the desired change flicker briefly and then disappear).  Looking at the ddp-analyzer result coming back, everything seems to be correct.  The trouble is that sometimes everything works correctly and sometimes it doesn't -- I can't reproduce the issue at will.  Do you want to have a poke around in our production environment at some stage to really see what's going on?  If so, email me and we'll set something up.

**Comment by arunoda** [comment]
Thanks for the update. I've a quite complex app in production, I'll try to
see if I can reproduce this.

On Mon, Dec 9, 2013 at 9:08 AM, Brent Abrahams notifications@github.comwrote:

> Updated to 0.3.24 and we're still noticing this quite regularly -- where
> the minimongo colllection is updated during latency compensation, the
> database collection is updated correctly on the server, then when the ddp
> result comes back to the client who initiated the change, the changes are
> undone on the client (net result as a user -- you see the desired change
> flicker briefly and then disappear). Looking at the ddp-analyzer result
> coming back, everything seems to be correct. The trouble is that sometimes
> everything works correctly and sometimes it doesn't -- I can't reproduce
> the issue at will. Do you want to have a poke around in our production
> environment at some stage to really see what's going on? If so, email me
> and we'll set something up.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/arunoda/meteor-smart-collections/issues/39#issuecomment-30103736
> .

## 

Arunoda Susiripala

@arunoda http://twitter.com/arunoda
http://gplus.to/arunodahttps://github.com/arunoda
http://www.linkedin.com/in/arunoda

**Comment by arunoda** [comment]
I've added another possible fix. check you please check with version: `v0.3.25`

**Comment by JackAdams** [comment]
Just observed the same issue a couple of times this morning.  The update gets reverted on the minimongo store on the client that initiated the change (once the DDP result has returned) but the database record is updated correctly and so are all other connected clients once the changes are pushed to them via DDP.  After this, refreshing the browser on the original client will then show the correctly updated data and everything will start working as normal again.
